File: src/detector.py
"""
Real-Time Object Detection Pipeline for Industrial Smart Environment Monitoring
Author: Monisha Ravi Kumar
Description:
    Implements a configurable YOLOv8-based detection pipeline supporting
    webcam, video file, and image directory inputs. Logs per-frame inference
    metrics (FPS, latency, class counts) and exports a structured JSON/CSV
    benchmark report for quantitative analysis.
"""

import cv2
import time
import json
import csv
import argparse
import numpy as np
from pathlib import Path
from datetime import datetime
from collections import defaultdict

try:
    from ultralytics import YOLO
    YOLO_AVAILABLE = True
except ImportError:
    YOLO_AVAILABLE = False


# ── Industrial-relevant COCO classes (subset) ──────────────────────────────
INDUSTRIAL_CLASSES = {
    0: "person",       # worker / personnel
    2: "car",          # forklift / vehicle (proxy)
    3: "motorcycle",   # mobility device
    7: "truck",        # heavy vehicle
    24: "backpack",    # personal equipment
    26: "handbag",     # carried object
    39: "bottle",      # container
    41: "cup",         # container
    56: "chair",       # workstation furniture
    57: "couch",       # rest area
    58: "potted plant",
    60: "dining table",# work surface (proxy)
    63: "laptop",      # control station
    64: "mouse",
    66: "keyboard",
    67: "cell phone",
    73: "book",        # documentation
    76: "scissors",    # tool (proxy)
    79: "toothbrush",  # small tool (proxy)
}


class IndustrialDetector:
    """
    YOLOv8 detection pipeline with per-frame benchmarking and
    structured result logging for industrial environment monitoring.
    """

    def __init__(self, model_size: str = "n", conf_threshold: float = 0.35,
                 iou_threshold: float = 0.45, output_dir: str = "results"):
        self.conf = conf_threshold
        self.iou = iou_threshold
        self.output_dir = Path(output_dir)
        self.output_dir.mkdir(parents=True, exist_ok=True)
        self.model_size = model_size

        # Metrics storage
        self.frame_metrics = []
        self.class_counts = defaultdict(int)
        self.session_id = datetime.now().strftime("%Y%m%d_%H%M%S")

        print(f"[INFO] Initialising IndustrialDetector (YOLOv8{model_size})")
        print(f"[INFO] Confidence threshold : {conf_threshold}")
        print(f"[INFO] IoU threshold        : {iou_threshold}")
        print(f"[INFO] Results directory    : {self.output_dir}")

        if YOLO_AVAILABLE:
            self.model = YOLO(f"yolov8{model_size}.pt")
            print(f"[INFO] Model loaded: yolov8{model_size}.pt")
        else:
            self.model = None
            print("[WARN] ultralytics not available — running in DEMO mode")

    # ── Core inference ─────────────────────────────────────────────────────

    def _infer(self, frame: np.ndarray):
        """Run YOLOv8 inference and return (results, latency_ms)."""
        t0 = time.perf_counter()
        if self.model:
            results = self.model(frame, conf=self.conf, iou=self.iou,
                                 verbose=False)[0]
        else:
            results = None
        latency = (time.perf_counter() - t0) * 1000
        return results, latency

    def _draw(self, frame: np.ndarray, results, fps: float) -> np.ndarray:
        """Draw bounding boxes, labels, and HUD onto frame."""
        overlay = frame.copy()

        if results and results.boxes is not None:
            for box in results.boxes:
                cls_id = int(box.cls[0])
                conf   = float(box.conf[0])
                label  = results.names.get(cls_id, str(cls_id))
                x1, y1, x2, y2 = map(int, box.xyxy[0])

                # Colour by class (deterministic)
                colour = self._class_colour(cls_id)
                cv2.rectangle(overlay, (x1, y1), (x2, y2), colour, 2)
                tag = f"{label} {conf:.2f}"
                (tw, th), _ = cv2.getTextSize(tag, cv2.FONT_HERSHEY_SIMPLEX, 0.55, 1)
                cv2.rectangle(overlay, (x1, y1 - th - 8), (x1 + tw + 4, y1), colour, -1)
                cv2.putText(overlay, tag, (x1 + 2, y1 - 4),
                            cv2.FONT_HERSHEY_SIMPLEX, 0.55, (255, 255, 255), 1)

        # HUD
        hud = f"FPS: {fps:5.1f}  |  YOLOv8{self.model_size}  |  conf={self.conf}"
        cv2.rectangle(overlay, (0, 0), (len(hud) * 9 + 10, 28), (20, 20, 20), -1)
        cv2.putText(overlay, hud, (6, 20),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.55, (0, 255, 120), 1)

        n_det = len(results.boxes) if (results and results.boxes is not None) else 0
        cv2.putText(overlay, f"Detections: {n_det}", (6, 48),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.55, (0, 200, 255), 1)
        return overlay

    @staticmethod
    def _class_colour(cls_id: int):
        np.random.seed(cls_id * 7 + 13)
        return tuple(int(c) for c in np.random.randint(80, 230, 3))

    def _log_frame(self, frame_idx: int, latency: float, fps: float, results):
        """Accumulate per-frame metrics."""
        det_classes = {}
        if results and results.boxes is not None:
            for box in results.boxes:
                name = results.names.get(int(box.cls[0]), "unknown")
                det_classes[name] = det_classes.get(name, 0) + 1
                self.class_counts[name] += 1

        self.frame_metrics.append({
            "frame":      frame_idx,
            "latency_ms": round(latency, 2),
            "fps":        round(fps, 2),
            "n_detections": sum(det_classes.values()),
            "classes":    det_classes,
        })

    # ── Input modes ────────────────────────────────────────────────────────

    def run_video(self, source, show: bool = True, save: bool = True,
                  max_frames: int = 0):
        """Process video file or webcam (source=0)."""
        cap = cv2.VideoCapture(source)
        if not cap.isOpened():
            raise RuntimeError(f"Cannot open source: {source}")

        w  = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        h  = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        src_fps = cap.get(cv2.CAP_PROP_FPS) or 30

        writer = None
        if save:
            out_path = self.output_dir / f"output_{self.session_id}.mp4"
            writer = cv2.VideoWriter(str(out_path),
                                     cv2.VideoWriter_fourcc(*"mp4v"),
                                     src_fps, (w, h))
            print(f"[INFO] Saving output to: {out_path}")

        frame_idx = 0
        fps_smooth = 0.0
        t_prev = time.perf_counter()

        print("[INFO] Running — press Q to quit")
        while True:
            ret, frame = cap.read()
            if not ret:
                break
            if max_frames and frame_idx >= max_frames:
                break

            results, latency = self._infer(frame)

            t_now   = time.perf_counter()
            fps_raw = 1.0 / max(t_now - t_prev, 1e-6)
            fps_smooth = 0.8 * fps_smooth + 0.2 * fps_raw  # EMA
            t_prev  = t_now

            vis = self._draw(frame, results, fps_smooth)
            self._log_frame(frame_idx, latency, fps_smooth, results)

            if writer:
                writer.write(vis)
            if show:
                cv2.imshow("Smart Factory Monitor", vis)
                if cv2.waitKey(1) & 0xFF == ord("q"):
                    break

            frame_idx += 1
            if frame_idx % 30 == 0:
                print(f"  frame {frame_idx:5d}  |  fps={fps_smooth:5.1f}"
                      f"  |  lat={latency:6.1f}ms")

        cap.release()
        if writer:
            writer.release()
        cv2.destroyAllWindows()
        print(f"[INFO] Processed {frame_idx} frames")
        self._export_results()

    def run_images(self, image_dir: str, show: bool = False):
        """Process a directory of images."""
        img_dir = Path(image_dir)
        paths = sorted(list(img_dir.glob("*.jpg")) +
                       list(img_dir.glob("*.png")) +
                       list(img_dir.glob("*.jpeg")))
        if not paths:
            print(f"[WARN] No images found in {image_dir}")
            return

        print(f"[INFO] Processing {len(paths)} images from {image_dir}")
        for idx, p in enumerate(paths):
            frame = cv2.imread(str(p))
            if frame is None:
                continue
            results, latency = self._infer(frame)
            fps = 1000.0 / max(latency, 1e-3)
            vis = self._draw(frame, results, fps)
            self._log_frame(idx, latency, fps, results)

            out_img = self.output_dir / f"det_{p.stem}.jpg"
            cv2.imwrite(str(out_img), vis)
            if show:
                cv2.imshow("Detection", vis)
                cv2.waitKey(500)

            print(f"  [{idx+1:3d}/{len(paths)}] {p.name}  lat={latency:.1f}ms")

        cv2.destroyAllWindows()
        self._export_results()

    # ── Export ─────────────────────────────────────────────────────────────

    def _export_results(self):
        """Export benchmark CSV, JSON summary, and print report."""
        if not self.frame_metrics:
            print("[WARN] No metrics to export.")
            return

        latencies = [m["latency_ms"] for m in self.frame_metrics]
        fpss      = [m["fps"]        for m in self.frame_metrics]
        dets      = [m["n_detections"] for m in self.frame_metrics]

        summary = {
            "session_id":       self.session_id,
            "model":            f"YOLOv8{self.model_size}",
            "conf_threshold":   self.conf,
            "iou_threshold":    self.iou,
            "total_frames":     len(self.frame_metrics),
            "avg_fps":          round(np.mean(fpss), 2),
            "min_fps":          round(np.min(fpss), 2),
            "max_fps":          round(np.max(fpss), 2),
            "avg_latency_ms":   round(np.mean(latencies), 2),
            "p95_latency_ms":   round(np.percentile(latencies, 95), 2),
            "p99_latency_ms":   round(np.percentile(latencies, 99), 2),
            "avg_detections_per_frame": round(np.mean(dets), 2),
            "total_detections": int(np.sum(dets)),
            "class_totals":     dict(self.class_counts),
        }

        # JSON
        json_path = self.output_dir / f"benchmark_{self.session_id}.json"
        with open(json_path, "w") as f:
            json.dump(summary, f, indent=2)

        # CSV (per-frame)
        csv_path = self.output_dir / f"frame_metrics_{self.session_id}.csv"
        with open(csv_path, "w", newline="") as f:
            writer = csv.DictWriter(f, fieldnames=["frame", "latency_ms",
                                                    "fps", "n_detections"])
            writer.writeheader()
            for m in self.frame_metrics:
                writer.writerow({k: m[k] for k in ["frame", "latency_ms",
                                                     "fps", "n_detections"]})

        # Console report
        print("\n" + "="*55)
        print("  BENCHMARK REPORT — Smart Factory Detection Pipeline")
        print("="*55)
        print(f"  Model          : YOLOv8{self.model_size}")
        print(f"  Frames         : {summary['total_frames']}")
        print(f"  Avg FPS        : {summary['avg_fps']}")
        print(f"  Avg latency    : {summary['avg_latency_ms']} ms")
        print(f"  P95 latency    : {summary['p95_latency_ms']} ms")
        print(f"  P99 latency    : {summary['p99_latency_ms']} ms")
        print(f"  Total detects  : {summary['total_detections']}")
        print(f"  Top classes    :")
        for cls, cnt in sorted(summary["class_totals"].items(),
                                key=lambda x: -x[1])[:5]:
            print(f"    {cls:<20s} {cnt}")
        print("="*55)
        print(f"  JSON  → {json_path}")
        print(f"  CSV   → {csv_path}")
        print("="*55 + "\n")
        return summary


# ── CLI ────────────────────────────────────────────────────────────────────

def parse_args():
    p = argparse.ArgumentParser(
        description="Smart Factory Real-Time Object Detection Pipeline")
    p.add_argument("--source", default="0",
                   help="Input: 0=webcam, path to video file, or image dir")
    p.add_argument("--model", default="n", choices=["n","s","m","l","x"],
                   help="YOLOv8 model size (n=nano … x=xlarge)")
    p.add_argument("--conf", type=float, default=0.35,
                   help="Confidence threshold (default: 0.35)")
    p.add_argument("--iou", type=float, default=0.45,
                   help="IoU NMS threshold (default: 0.45)")
    p.add_argument("--output", default="results",
                   help="Output directory for results")
    p.add_argument("--no-show", action="store_true",
                   help="Disable live display (headless mode)")
    p.add_argument("--no-save", action="store_true",
                   help="Disable video saving")
    p.add_argument("--max-frames", type=int, default=0,
                   help="Max frames to process (0=all)")
    return p.parse_args()


if __name__ == "__main__":
    args = parse_args()
    detector = IndustrialDetector(
        model_size=args.model,
        conf_threshold=args.conf,
        iou_threshold=args.iou,
        output_dir=args.output,
    )

    src = args.source
    if src.isdigit():
        src = int(src)
        detector.run_video(src, show=not args.no_show,
                           save=not args.no_save,
                           max_frames=args.max_frames)
    elif Path(src).is_dir():
        detector.run_images(src, show=not args.no_show)
    else:
        detector.run_video(src, show=not args.no_show,
                           save=not args.no_save,
                           max_frames=args.max_frames)

File: cpp/detector.cpp
/**
 * Smart Factory Object Detection — C++ Inference Engine
 * Author: Monisha Ravi Kumar
 *
 * Loads a YOLOv8 ONNX model via OpenCV DNN module and runs
 * real-time inference on webcam / video / image input.
 *
 * Build:
 *   g++ -std=c++17 -O2 detector.cpp \
 *       $(pkg-config --cflags --libs opencv4) \
 *       -o smart_factory_detector
 *
 * Run:
 *   ./smart_factory_detector --model yolov8n.onnx --source 0
 *   ./smart_factory_detector --model yolov8n.onnx --source factory.mp4
 */

#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <chrono>
#include <numeric>
#include <algorithm>
#include <iomanip>
#include <sstream>

// ── Configuration ──────────────────────────────────────────────────────────

struct Config {
    std::string model_path  = "yolov8n.onnx";
    std::string source      = "0";
    std::string output_dir  = "results";
    float  conf_threshold   = 0.35f;
    float  nms_threshold    = 0.45f;
    int    input_width      = 640;
    int    input_height     = 640;
    bool   show_window      = true;
    bool   save_video       = true;
    int    max_frames       = 0;      // 0 = unlimited
};

// COCO class names (80 classes)
const std::vector<std::string> CLASS_NAMES = {
    "person","bicycle","car","motorcycle","airplane","bus","train","truck",
    "boat","traffic light","fire hydrant","stop sign","parking meter","bench",
    "bird","cat","dog","horse","sheep","cow","elephant","bear","zebra","giraffe",
    "backpack","umbrella","handbag","tie","suitcase","frisbee","skis","snowboard",
    "sports ball","kite","baseball bat","baseball glove","skateboard","surfboard",
    "tennis racket","bottle","wine glass","cup","fork","knife","spoon","bowl",
    "banana","apple","sandwich","orange","broccoli","carrot","hot dog","pizza",
    "donut","cake","chair","couch","potted plant","bed","dining table","toilet",
    "tv","laptop","mouse","remote","keyboard","cell phone","microwave","oven",
    "toaster","sink","refrigerator","book","clock","vase","scissors","teddy bear",
    "hair drier","toothbrush"
};

// ── Utility ────────────────────────────────────────────────────────────────

cv::Scalar classColour(int cls_id) {
    // Deterministic per-class colour
    cv::RNG rng(cls_id * 7 + 13);
    return cv::Scalar(rng.uniform(80, 230),
                      rng.uniform(80, 230),
                      rng.uniform(80, 230));
}

std::string formatMs(double ms) {
    std::ostringstream ss;
    ss << std::fixed << std::setprecision(1) << ms << " ms";
    return ss.str();
}

// ── Detection result ───────────────────────────────────────────────────────

struct Detection {
    int   class_id;
    float confidence;
    cv::Rect box;
};

// ── YOLOv8 post-processing ─────────────────────────────────────────────────

/**
 * YOLOv8 ONNX output shape: [1, 84, 8400]
 * Rows 0-3  : cx, cy, w, h
 * Rows 4-83 : class scores (80 classes)
 */
std::vector<Detection> postprocess(const cv::Mat& output,
                                   float conf_thresh,
                                   float nms_thresh,
                                   int orig_w, int orig_h,
                                   int input_w, int input_h) {
    std::vector<cv::Rect>  boxes;
    std::vector<float>     scores;
    std::vector<int>       class_ids;

    // output shape: [1, 84, 8400] → reshape to [84, 8400]
    cv::Mat out = output.reshape(1, output.size[1]);  // 84 x 8400

    float scale_x = (float)orig_w / input_w;
    float scale_y = (float)orig_h / input_h;

    for (int i = 0; i < out.cols; ++i) {
        // Find max class score
        cv::Mat class_scores = out.col(i).rowRange(4, out.rows);
        double max_score;
        cv::Point max_loc;
        cv::minMaxLoc(class_scores, nullptr, &max_score, nullptr, &max_loc);

        if ((float)max_score < conf_thresh) continue;

        float cx = out.at<float>(0, i) * scale_x;
        float cy = out.at<float>(1, i) * scale_y;
        float w  = out.at<float>(2, i) * scale_x;
        float h  = out.at<float>(3, i) * scale_y;

        int x1 = std::max(0, (int)(cx - w / 2));
        int y1 = std::max(0, (int)(cy - h / 2));
        int bw = std::min((int)w, orig_w - x1);
        int bh = std::min((int)h, orig_h - y1);

        boxes.push_back(cv::Rect(x1, y1, bw, bh));
        scores.push_back((float)max_score);
        class_ids.push_back(max_loc.y);
    }

    // NMS
    std::vector<int> indices;
    cv::dnn::NMSBoxes(boxes, scores, conf_thresh, nms_thresh, indices);

    std::vector<Detection> detections;
    for (int idx : indices) {
        detections.push_back({class_ids[idx], scores[idx], boxes[idx]});
    }
    return detections;
}

// ── Drawing ────────────────────────────────────────────────────────────────

void drawDetections(cv::Mat& frame,
                    const std::vector<Detection>& dets,
                    double fps, const std::string& model_name) {
    for (const auto& d : dets) {
        auto colour = classColour(d.class_id);
        cv::rectangle(frame, d.box, colour, 2);

        std::string label = CLASS_NAMES[d.class_id] + " "
                          + std::to_string((int)(d.confidence * 100)) + "%";
        int baseline = 0;
        auto sz = cv::getTextSize(label, cv::FONT_HERSHEY_SIMPLEX,
                                   0.55, 1, &baseline);
        cv::rectangle(frame,
                      cv::Point(d.box.x, d.box.y - sz.height - 8),
                      cv::Point(d.box.x + sz.width + 4, d.box.y),
                      colour, cv::FILLED);
        cv::putText(frame, label,
                    cv::Point(d.box.x + 2, d.box.y - 4),
                    cv::FONT_HERSHEY_SIMPLEX, 0.55,
                    cv::Scalar(255, 255, 255), 1);
    }

    // HUD
    std::ostringstream hud;
    hud << "FPS: " << std::fixed << std::setprecision(1) << fps
        << "  |  " << model_name << "  |  dets=" << dets.size();
    cv::rectangle(frame, cv::Point(0, 0),
                  cv::Point((int)hud.str().size() * 9 + 10, 28),
                  cv::Scalar(20, 20, 20), cv::FILLED);
    cv::putText(frame, hud.str(), cv::Point(6, 20),
                cv::FONT_HERSHEY_SIMPLEX, 0.55,
                cv::Scalar(0, 255, 120), 1);
}

// ── Benchmark report ───────────────────────────────────────────────────────

void printReport(const std::string& model_name,
                 const std::vector<double>& latencies) {
    if (latencies.empty()) return;

    auto sorted = latencies;
    std::sort(sorted.begin(), sorted.end());

    double avg = std::accumulate(sorted.begin(), sorted.end(), 0.0)
               / sorted.size();
    double p95 = sorted[(size_t)(sorted.size() * 0.95)];
    double p99 = sorted[(size_t)(sorted.size() * 0.99)];
    double fps  = 1000.0 / avg;

    std::cout << "\n" << std::string(55, '=') << "\n";
    std::cout << "  BENCHMARK REPORT — C++ Inference Engine\n";
    std::cout << std::string(55, '=') << "\n";
    std::cout << "  Model        : " << model_name       << "\n";
    std::cout << "  Frames       : " << latencies.size() << "\n";
    std::cout << "  Avg FPS      : " << std::fixed << std::setprecision(1) << fps  << "\n";
    std::cout << "  Avg latency  : " << formatMs(avg) << "\n";
    std::cout << "  P95 latency  : " << formatMs(p95) << "\n";
    std::cout << "  P99 latency  : " << formatMs(p99) << "\n";
    std::cout << "  Min latency  : " << formatMs(sorted.front()) << "\n";
    std::cout << "  Max latency  : " << formatMs(sorted.back())  << "\n";
    std::cout << std::string(55, '=') << "\n\n";
}

// ── Main ───────────────────────────────────────────────────────────────────

int main(int argc, char** argv) {
    Config cfg;

    // Simple CLI parsing
    for (int i = 1; i < argc; ++i) {
        std::string arg(argv[i]);
        if      (arg == "--model"  && i+1 < argc) cfg.model_path    = argv[++i];
        else if (arg == "--source" && i+1 < argc) cfg.source        = argv[++i];
        else if (arg == "--conf"   && i+1 < argc) cfg.conf_threshold= std::stof(argv[++i]);
        else if (arg == "--output" && i+1 < argc) cfg.output_dir    = argv[++i];
        else if (arg == "--no-show")               cfg.show_window   = false;
        else if (arg == "--no-save")               cfg.save_video    = false;
        else if (arg == "--max-frames" && i+1 < argc) cfg.max_frames = std::stoi(argv[++i]);
    }

    std::cout << "[INFO] Smart Factory Detection Pipeline (C++ / OpenCV DNN)\n";
    std::cout << "[INFO] Model  : " << cfg.model_path << "\n";
    std::cout << "[INFO] Source : " << cfg.source     << "\n";
    std::cout << "[INFO] Conf   : " << cfg.conf_threshold << "\n\n";

    // Load model
    cv::dnn::Net net;
    try {
        net = cv::dnn::readNetFromONNX(cfg.model_path);
    } catch (const cv::Exception& e) {
        std::cerr << "[ERROR] Failed to load model: " << e.what() << "\n";
        std::cerr << "        Export with: yolo export model=yolov8n.pt format=onnx\n";
        return 1;
    }
    net.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
    net.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);

    // Open source
    cv::VideoCapture cap;
    bool is_webcam = (cfg.source == "0");
    if (is_webcam) cap.open(0);
    else           cap.open(cfg.source);

    if (!cap.isOpened()) {
        std::cerr << "[ERROR] Cannot open source: " << cfg.source << "\n";
        return 1;
    }

    int orig_w = (int)cap.get(cv::CAP_PROP_FRAME_WIDTH);
    int orig_h = (int)cap.get(cv::CAP_PROP_FRAME_HEIGHT);
    double src_fps = cap.get(cv::CAP_PROP_FPS);
    if (src_fps <= 0) src_fps = 30.0;

    std::cout << "[INFO] Source resolution: " << orig_w << "x" << orig_h
              << "  @" << src_fps << " fps\n";

    // Video writer
    cv::VideoWriter writer;
    if (cfg.save_video) {
        std::string out_path = cfg.output_dir + "/output_cpp.mp4";
        writer.open(out_path,
                    cv::VideoWriter::fourcc('m','p','4','v'),
                    src_fps, cv::Size(orig_w, orig_h));
        std::cout << "[INFO] Saving to: " << out_path << "\n";
    }

    std::vector<double> latencies;
    double fps_smooth = 0.0;
    auto t_prev = std::chrono::steady_clock::now();
    int frame_idx = 0;

    std::cout << "[INFO] Running — press Q to quit\n\n";

    while (true) {
        cv::Mat frame;
        if (!cap.read(frame) || frame.empty()) break;
        if (cfg.max_frames > 0 && frame_idx >= cfg.max_frames) break;

        // Pre-process: letterbox → blob
        cv::Mat blob;
        cv::dnn::blobFromImage(frame, blob, 1.0 / 255.0,
                               cv::Size(cfg.input_width, cfg.input_height),
                               cv::Scalar(), true, false);

        // Inference
        auto t0 = std::chrono::steady_clock::now();
        net.setInput(blob);
        std::vector<cv::Mat> outputs;
        net.forward(outputs, net.getUnconnectedOutLayersNames());
        auto t1 = std::chrono::steady_clock::now();

        double latency = std::chrono::duration<double, std::milli>(t1 - t0).count();
        latencies.push_back(latency);

        // FPS (EMA)
        auto t_now = std::chrono::steady_clock::now();
        double raw_fps = 1e3 / std::chrono::duration<double, std::milli>(
                                   t_now - t_prev).count();
        fps_smooth = 0.8 * fps_smooth + 0.2 * raw_fps;
        t_prev = t_now;

        // Post-process
        auto dets = postprocess(outputs[0], cfg.conf_threshold,
                                cfg.nms_threshold,
                                orig_w, orig_h,
                                cfg.input_width, cfg.input_height);

        // Draw
        drawDetections(frame, dets, fps_smooth, "YOLOv8-ONNX");

        if (cfg.save_video && writer.isOpened()) writer.write(frame);
        if (cfg.show_window) {
            cv::imshow("Smart Factory Monitor (C++)", frame);
            if ((cv::waitKey(1) & 0xFF) == 'q') break;
        }

        frame_idx++;
        if (frame_idx % 30 == 0) {
            std::cout << "  frame " << std::setw(5) << frame_idx
                      << "  |  fps=" << std::fixed << std::setprecision(1)
                      << fps_smooth
                      << "  |  lat=" << formatMs(latency)
                      << "  |  dets=" << dets.size() << "\n";
        }
    }

    cap.release();
    cv::destroyAllWindows();

    printReport(cfg.model_path, latencies);
    return 0;
}
